### PR TITLE
add redeem_allowed to protect excess withdraw

### PR DESCRIPTION
**Pull request by yz89**

### CODE REVIEW

**Comment by GopherJ** [comment]
I think:

```
ensure_borrow_allowed
ensure_redeem_allowed
```

will be better, not sure how the others think

**Review by GopherJ — APPROVED** [review]

File: pallets/loans/src/lib.rs [hunk 1]
```diff
@@ -833,37 +835,22 @@ impl<T: Config> Pallet<T> {
...
-
-        Ok(())
+    fn get_account_liquidity(
+        account: &T::AccountId,
+    ) -> Result<(FixedU128, FixedU128), DispatchError> {
+        let total_borrow_value = Self::total_borrowed_value(account)?;
+        let total_collateral_value = Self::total_collateral_value(account)?;
+        if total_collateral_value > total_borrow_value {
+            Ok((
+                total_collateral_value - total_borrow_value,
```
Comment (c410-f3r) [inline pallets/loans/src/lib.rs:845]: This line probably won't overflow in practice but maybe play safe and use `total_collateral_value.checked_sub(total_borrow_value).or_ok(ArithmeticError::Underflow)?` instead?

**Review by c410-f3r — APPROVED** [review]

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -940,17 +909,85 @@ impl<T: Config> Pallet<T> {
...
+            return Err(Error::<T>::InsufficientDeposit.into());
+        }
+        if !deposit.is_collateral {
+            return Ok(());
+        }
+        let collateral_factor = Self::collateral_factor(currency_id);
+        let price = Self::get_price(currency_id)?;
+        let redeem_effects_value = price
+            .checked_mul(&FixedU128::from_inner(
+                collateral_factor.mul_ceil(redeem_amount),
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:931]: why do we use `mul_ceil` here?

Comment (GopherJ) [inline pallets/loans/src/lib.rs:931]: 👍

File: pallets/loans/src/lib.rs (hunk 1 above)
Comment (yz89) [reply pallets/loans/src/lib.rs:845]: I don't think it's necessary for It's an easy case.

**Pull request merged**